### PR TITLE
[WIP] MP3 support

### PR DESCRIPTION
**Pull request by foolswood**
Putting this up early (in a distinctly unpolished state) to get some feedback about the overall shape and approach.

Status summary:
- [x] Read support
- [ ] Write support
- [ ] Make optional in build
- [x] dlopen lame

### CODE REVIEW

**Comment by erikd** [comment]
> however in order to test it an example file to load is required. Is this acceptable, and if so where should this be in the source tree?

libsndfile has been around for nearly 20 years and has *never* had a binary file committed to the repository. i would like to keep it that way :). Hence, basics of read and write should be added in one go.

**Comment by foolswood** [comment]
Fair enough, I'll have a stab at that. Did you have any comments on what's there so far?

File: src/mp3.c [hunk 1]
```diff
@@ -227,16 +241,72 @@ mp3_read_2d	(SF_PRIVATE * psf, double * ptr, sf_count_t len)
...
+	char * err = NULL ;
+	while (lame_handle.dlhandle == NULL)
+	{	lame_handle.dlhandle = dlopen ("liblame.so", RTLD_LAZY) ;
+		if (lame_handle.dlhandle == NULL)
+			return dlerror () ;
+		lame_handle.init = dlsym (lame_handle.dlhandle, "lame_init") ;
+		if (lame_handle.init == NULL)
+		{	failed = 1 ;
+			break ;
+		}
```
Comment (erikd) [inline src/mp3.c]: Those braces don't match the coding style of the rest of the libsndfile code base,

The `Scripts/cstyle.py` script catches some, but not all issue.

File: src/mp3.c [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+mp3_open (SF_PRIVATE * psf)
+{	psf->codec_data = NULL ;
+	switch (psf->file.mode) {
+	case (SFM_READ) :
+		return mp3_read_open (psf) ;
+	case (SFM_WRITE) :
+		return mp3_write_open (psf) ;
+	default:
+		return SFE_UNIMPLEMENTED ;
+	}
```
Comment (erikd) [inline src/mp3.c]: The libsndfile codebase has an coding style which it adheres to with a reasonably high level of consistency.

The is an a Python script `Scripts/cstyle.py` which catchs a large number of issues. You should still have a look at the other code and match the style.

File: src/mp3.c [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+*/
+
+// FIXME: use mpg123 error string reporting
+static int
+mp3_read_open (SF_PRIVATE * psf)
+{
+	int decoder_err = MPG123_OK ;
+	mpg123_handle * decoder ;
+	if (mpg123_initialised == 0)
+	{	int decoder_init_err = mpg123_init () ;
```
Comment (erikd) [inline src/mp3.c]: Can `mpg123_init()` not be called multiple times without blowing up?

File: src/mp3.c [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+	psf->read_double = mp3_read_2d ;
+
+	return 0 ;
+}
+
+static int
+mp3_read_close (SF_PRIVATE * psf)
+{	mpg123_handle * decoder = psf->codec_data ;
+	if (decoder != NULL)
+	{	// mpg123_close(decoder); <- Not sure if we need this?
```
Comment (erikd) [inline src/mp3.c]: If it doesn't blow up, call it anyway.

File: src/mp3.c [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+		mpg123_delete (decoder) ;
+		// The psf sometimes gets reused:
+		psf->codec_data = NULL ;
+	}
+	if (!--mpg123_initialised)
+		mpg123_exit () ;
```
Comment (erikd) [inline src/mp3.c]: This is kinda weird! What happens without that `mpg123_exit()` call? Does it leak memory?

File: src/mp3.c (hunk 4 above)
Comment (foolswood) [reply src/mp3.c]: IIRC it did, which is why it was commented out, but I'll check.

File: src/mp3.c (hunk 2 above)
Comment (foolswood) [reply src/mp3.c]: I thought I had got it all past cstyle.py (which I believe is triggered in the pre-commit hook), but I'll check elsewhere for the preferred bracket style.

File: src/mp3.c (hunk 5 above)
Comment (foolswood) [reply src/mp3.c]: The `mpg123_init()` and `mpg123_exit()` calls are following the canonical documented use of the library. I've looked through the source, which suggests there is some cheating to be had and contacted the mpg123 developers to see what their future plans for the methods are.

File: src/mp3.c (hunk 1 above)
Comment (erikd) [reply src/mp3.c]: I've just added some new rules to the `Script/cstyle.py` script that should help with some more coding style issues. The one that's difficult to add to that script is the closing curly indentation which the current source code uses. It should be like:
```
    if (whatever)
    {   do_soometing ;
        } ;
```
Notice how the closing curly is indented one more level than the open curly brace.

**Comment by foolswood** [comment]
@diemoschwarz I had dispaired of this somewhat (see the discussion here and on #258 for why). IIRC (which I might not because this was 18 months ago) I did have playback working at one point (although I can't remember under what conditions).

**Comment by evpobr** [comment]
#499 is almost ready, closing this one.

**Pull request closed without merging**